### PR TITLE
Rendering materials

### PR DESCRIPTION
**Pull request by Connor-Colenso**
Adds item rendering to the two Avaritia materials and introduces a new material called Universium.

https://user-images.githubusercontent.com/52056774/215242229-851d23b2-96a5-4b4c-bab4-b345de557a21.mov

### CODE REVIEW

File: src/main/java/gregtech/common/render/items/InfinityRenderer.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        t.startDrawingQuads();
+        t.addVertexWithUV(-spread, -spread, 0, halo.getMinU(), halo.getMinV());
+        t.addVertexWithUV(-spread, 16 + spread, 0, halo.getMinU(), halo.getMaxV());
+        t.addVertexWithUV(16 + spread, 16 + spread, 0, halo.getMaxU(), halo.getMaxV());
+        t.addVertexWithUV(16 + spread, -spread, 0, halo.getMaxU(), halo.getMinV());
+        t.draw();
+
+        GL11.glPushMatrix();
+        double scale = (rand.nextGaussian() * 0.15) + 0.95;
```
Comment (eigenraven) [inline src/main/java/gregtech/common/render/items/InfinityRenderer.java:56]: It's a bad idea to randomize things like this per frame, use a noise function parametrized by time instead for a smoothed change that doesn't depend on the framerate on your computer (see <https://iquilezles.org/articles/gradientnoise/>, those are 2d but the same idea exists in 1d)

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java [hunk 2]
```diff
@@ -17434,7 +17434,7 @@ private void addRecipesApril2017ChemistryUpdate() {
         // Custom Sodium Persulfate Ore Processing Recipes
         GT_Values.RA.addChemicalBathRecipe(
                 GT_OreDictUnificator.get(OrePrefixes.crushed, Materials.Tantalite, 1),
-                Materials.SodiumPersulfate.getFluid(500L),
+                Materials.SodiumPersulfate.getFluid(100L),
```
Comment (eigenraven) [inline src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:17437]: Did you include the changes to this file by accident? They seem unrelated

**Review by eigenraven — DISMISSED** [review]

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_BasicBatteryBuffer.java [hunk 3]
```diff
@@ -376,16 +378,60 @@ public boolean useModularUI() {
     public void addUIWidgets(ModularWindow.Builder builder, UIBuildContext buildContext) {
         switch (mInventory.length) {
             case 4:
-                getBaseMetaTileEntity().add2by2Slots(builder);
+                builder.widget(SlotGroup.ofItemHandler(inventoryHandler, 2)
```
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_BasicBatteryBuffer.java:381]: merge shenanigans https://github.com/GTNewHorizons/GT5-Unofficial/pull/1590

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 2 above)
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:17437]: merge shenanigans https://github.com/GTNewHorizons/GT5-Unofficial/pull/1591

File: src/main/java/gregtech/common/render/items/GT_MetaGenerated_Item_Renderer.java [hunk 4]
```diff
@@ -58,6 +61,15 @@ private IItemRenderer getRendererForItemStack(ItemStack aStack) {
 
         if (aMetaData < aItem.mOffset) {
             IItemRenderer aMaterialRenderer = aItem.getMaterialRenderer(aMetaData);
+
+            // Handle fluid rendering.
+            if (aMaterialRenderer == null) {
+                try {
+                    Materials material = getAssociation(aStack).mMaterial.mMaterial;
+                    aMaterialRenderer = material.renderer;
+                } catch(Exception ignored) {}
```
Comment (Glease) [inline src/main/java/gregtech/common/render/items/GT_MetaGenerated_Item_Renderer.java:70]: This is rendering code. Catch null pointer using if, not try catch an NPE, as otherwise we will soon have mysterious lag randomly when scrolling through NEI.

File: src/main/java/gregtech/common/render/items/InfinityRenderer.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        IIcon tIcon = aIconContainer.getIcon();
+        IIcon tOverlay = aIconContainer.getOverlayIcon();
+        FluidStack aFluid = GT_Utility.getFluidForFilledItem(aStack, true);
+
+        GL11.glEnable(GL11.GL_BLEND);
+        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
+        GL11.glEnable(GL11.GL_ALPHA_TEST);
+
+
+        GL11.glDisable(GL11.GL_BLEND);
```
Comment (Glease) [inline src/main/java/gregtech/common/render/items/InfinityRenderer.java:51]: No point in enable then disable.

File: src/main/java/gregtech/common/render/items/InfinityRenderer.java (hunk 1 above)
Comment (Glease) [reply src/main/java/gregtech/common/render/items/InfinityRenderer.java:127]: FIY renderPartialTicks aren't passed here. Get it from `gregtech.common.GT_Client#getPartialRenderTicks`.

File: src/main/java/gregtech/common/render/items/UniversiumRenderer.java [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    public boolean handleRenderType(ItemStack item, ItemRenderType type) {
+        return true;
+    }
+
+    @Override
+    public boolean shouldUseRenderHelper(ItemRenderType type, ItemStack item, ItemRendererHelper helper) {
+        return helper == ItemRendererHelper.ENTITY_ROTATION || helper == ItemRendererHelper.ENTITY_BOBBING;
+    }
+
+//    @Override
```
Comment (Glease) [inline src/main/java/gregtech/common/render/items/UniversiumRenderer.java:40]: please remove commented out code

File: src/main/java/gregtech/common/render/items/UniversiumRenderer.java [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        }
+
+        IIcon tIcon = aIconContainer.getIcon();
+        IIcon tOverlay = aIconContainer.getOverlayIcon();
+
+        GL11.glEnable(GL11.GL_BLEND);
+        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
+        GL11.glEnable(GL11.GL_ALPHA_TEST);
+
+        if (tOverlay != null) {
```
Comment (Glease) [inline src/main/java/gregtech/common/render/items/UniversiumRenderer.java:63]: Why are you rendering an overlay first? This seems to go against the idea of "overlay"

File: src/main/java/gregtech/common/render/items/InfinityRenderer.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+        renderHalo();
+
+        if (tOverlay != null) {
```
Comment (Glease) [inline src/main/java/gregtech/common/render/items/InfinityRenderer.java:55]: Why are you rendering an overlay first? This seems to go against the idea of "overlay"

File: src/main/java/gregtech/common/render/items/UniversiumRenderer.java (hunk 7 above)
Comment (Connor-Colenso) [reply src/main/java/gregtech/common/render/items/UniversiumRenderer.java:63]: If I don't do it this way the item darkens. 
Screenshot 2023-01-28 at 04 06 15

Comment (eigenraven) [reply src/main/java/gregtech/common/render/items/UniversiumRenderer.java:63]: Rename it to underlay and the code will make more sense

Comment (Glease) [reply src/main/java/gregtech/common/render/items/UniversiumRenderer.java:63]: Sounds like you have put overlay at where normal icon is expected really

**Review by Dream-Master — APPROVED** [review]
